internals/overlord/tlsstate/manager.go

```diff
@@ -0,0 +1,511 @@
+// Copyright (c) 2025 Canonical Ltd
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 as
+// published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+// tlsstate manages TLS keypairs on behalf of the daemon's HTTPS server.
+package tlsstate
+
+import (
+	"bytes"
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"crypto/sha512"
+	"crypto/tls"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/base32"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"io/fs"
+	"math/big"
+	"os"
+	"path/filepath"
+	"regexp"
+	"slices"
+	"strings"
+	"sync"
+	"time"
+)
+
+var (
+	// systemTime can be faked during testing.
+	systemTime = time.Now
+
+	// certificateValidityPeriod determines how long a newly generated
+	// TLS keypair is valid for.
+	certificateValidityPeriod = (10 * 365 * 24 * time.Hour)
+)
+
+type TLSManager struct {
+	// tlsDir is the location of the PEM keypair files.
+	tlsDir         string
+	tlsKeyPairLock sync.Mutex
+	tlsKeyPairs    []*TLSKeyPair
+}
+
+// NewManager creates a new TLS manager and loads existing keypairs from the
+// tlsDir directory.
+//
+// The TLS manager will generate keypairs on demand, so no external keypair
+// generation or lifecycle management of existing keypairs will be required.
+//
+// However, is possible to pre-seed the tlsDir with TLSv1.3 supported keypairs.
+// The following steps provide an example of how to create a TLSv1.3 supported
+// self-signed TLS keypair (the private key must be encoded in PKCS8 format):
+//
+// ~/> openssl genpkey -algorithm EC -pkeyopt ec_paramgen_curve:prime256v1 -out ec-p256.pem
+// ~/> openssl req -new -x509 -key ec-p256.pem -out ec-p256-cert.pem -days 365
+//
+// The name of the file must match the SHA512/384 hash of the raw public certificate
+// encoded in base32 (without padding), and contain both the private and certificate
+// blocks in a single PEM file:
+//
+// ~/> FINGERPRINT=$(openssl x509 -in ec-p256-cert.pem -outform DER | openssl dgst -sha384 -binary | base32 -w0 | tr -d '=')
+// ~/> cat ec-p256.pem > ${FINGERPRINT}.pem
+// ~/> cat ec-p256-cert.pem >> ${FINGERPRINT}.pem
```

Comment:
Appending one file to the other sounds like a weird ad-hoc file format, which is neither a certificate nor a key and is not the expected content of a `.pem`. Is this a standard practice? If not, is there really a reason to do this instead of having two files: `${FINGERPRINT}.pem` and `${FINGERPRINT}-cert.pem` ?

Reply:
PEM is a multi block container format, so that part is not unusual.

We can split it in two, but I am not sure that makes the solution safer from a security standpoint. It does make the code significantly more complex (I had it like that before). I think the next step would be to move the private key into a hardware backed keystore, as part of a secure boot capable system.

I am happy to split it in two if that is the consensus.

Reply:
I've skimmed through https://stackoverflow.com/questions/5355046/where-is-the-pem-file-format-specified which suggests that just concatenating the two pieces of data is probably fine. And you're right that it makes the code simpler.